### PR TITLE
🏗  ♻️  📖  Add JSDoc to every function in /build-system

### PR DESCRIPTION
**Pull request by rileyajones**
There should be no functional changes in this PR.

### CODE REVIEW

File: build-system/tasks/helpers.js [hunk 1]
```diff
@@ -251,8 +279,9 @@ function combineWithCompiledFile(srcFilename, destFilePath, options) {
     file: destFileName,
     hires: true,
   });
+
   const remapped = remapping(
-    bundledMap,
+    {...bundledMap, version: 3},
```
Comment (danielrozenberg) [inline build-system/tasks/helpers.js:284]: Maybe split these changes into a separate PR?

File: build-system/.eslintrc.js [hunk 2]
```diff
@@ -48,6 +48,6 @@ module.exports = {
     'local/no-module-exports': 0,
     'local/no-rest': 0,
     'local/no-spread': 0,
-    'require-jsdoc': 0,
+    'require-jsdoc': 1,
```
Comment (danielrozenberg) [inline build-system/.eslintrc.js:51]: I strongly suggest that you make this change in a follow-up PR - when your PR edits so many files the odds of there being another change that edits one of the same files increases significantly, and it's quite possible that someone will add a new function or change an existing one without the required JSDoc - which would cause the two merged to result in a codebase that breaks the build (and on a mere technicality no less)

If you first merge the changes, then wait a ~day (or at least couple hours), and then make the change, you are much more likely to avoid this situation

File: build-system/tasks/helpers.js (hunk 1 above)
Comment (rileyajones) [reply build-system/tasks/helpers.js:284]: Makes sense

File: build-system/.eslintrc.js (hunk 2 above)
Comment (rileyajones) [reply build-system/.eslintrc.js]: That is a great point

Comment (rsimha) [reply build-system/.eslintrc.js]: +1. Thanks for this substantial cleanup @rileyajones, and thanks for pointing this out, @danielrozenberg!

File: build-system/babel-config/eslint-config.js [hunk 3]
```diff
@@ -37,9 +37,14 @@ function getEslintConfig() {
   };
 }
 
+/**
+ * @return {{
+ *  manipulateOptions: {Function(_opts: *, parserOpts: *): void}
+ * }}
```
Comment (rsimha) [inline build-system/babel-config/eslint-config.js:43]: TIL JSDoc types can be split across multiple lines like this. (In the past, I remember closure complaining about multiline types in runtime code. Probably not an issue when we start checking build system code with TS.)

**Review by danielrozenberg — APPROVED** [review]
LGTM by me but I'll suggest getting approval from at least one other person :)

File: build-system/babel-plugins/babel-plugin-transform-promise-resolve/index.js [hunk 4]
```diff
@@ -17,6 +17,17 @@
...
+ * @param {*} babel
+ * @param {{
+ *  importFrom?: string,
+ * }=} options
+ * @return {{
+ *   visitor: {
+ * CallExpression: {Funcion(path: string): void}
+ *  }
+ * }}
+ */
```
Comment (rsimha) [inline build-system/babel-plugins/babel-plugin-transform-promise-resolve/index.js:30]: General question for folks like @jridgewell, @samouri, and @erwinmombay who add / maintain most of our custom babel plugins. Is it too much of a tax to pay to keep these types accurate?

In a future PR, @rileyajones is going to make this type info a requirement, similar to runtime code. On the plus side, he has done the hard work of adding all missing types before enforcing the rule 😃

File: build-system/tasks/bundle-size/index.js [hunk 5]
```diff
@@ -207,6 +207,9 @@ async function reportBundleSize() {
   }
 }
 
+/**
+ * @return {Promise<void>}
+ */
```
Comment (rsimha) [inline build-system/tasks/bundle-size/index.js:212]: Re: this change, @samouri is gonna say _"I told you so"_. And I will sheepishly look away and whistle a tune.

File: build-system/tsconfig.json [hunk 6]
```diff
@@ -26,7 +26,16 @@
...
+    "./tasks/**/*.ts",
+    "./tasks/**/*.ts"
   ],
-  "exclude": ["**/node_modules", "common/check-package-manager.js"]
+  "exclude": [
+    "**/node_modules",
+    "common/check-package-manager.js",
+    "server/app-index/test/test-html.js",
+    "server/app-index/test/test-amphtml-helpers.js",
+    "server/new-server/transforms/dist",
```
Comment (rsimha) [inline build-system/tsconfig.json:38]: Maintenance nit: `**/node_modules` and `server/new-server/transforms/dist` are output directories that are not part of the codebase. Would it be possible to just ignore them while running checks instead of having to explicitly list them here? Would the change in #32960 help with this?

**Review by rsimha — APPROVED** [review]
Heroic effort. LGTM!

I've left some non-blocking comments below. Feel free to punt to later PRs if needed.

File: build-system/tasks/bundle-size/index.js (hunk 5 above)
Comment (samouri) [reply build-system/tasks/bundle-size/index.js:212]: ooh, the sweet sweet taste of being right on the internet

**Review by estherkim — APPROVED** [review]

File: build-system/babel-plugins/babel-plugin-transform-promise-resolve/index.js (hunk 4 above)
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-transform-promise-resolve/index.js]: They aren't so frequently updated, and keeping types correct is always a worthwhile trade

File: .gitignore [hunk 7]
```diff
@@ -13,6 +13,8 @@ npm-debug.log
 .idea
 .tm_properties
 .settings
+.karma-cache/
+test-bin/
```
Comment (rsimha) [inline .gitignore:17]: These are the result of a bad merge and should be removed.

File: build-system/tsconfig.json [hunk 8]
```diff
@@ -26,7 +26,9 @@
     "./common/**/*.js",
     "./common/**/*.ts",
     "./server/**/*.js",
-    "./server/**/*.ts"
+    "./server/**/*.ts",
+    "./tasks/**/*.ts",
+    "./tasks/**/*.ts"
```
Comment (rsimha) [inline build-system/tsconfig.json:31]: I'm guessing one of these should end with `.js`?

File: .gitignore (hunk 7 above)
Comment (rileyajones) [reply .gitignore:17]: I seriously thought I reverted this change. I'll clean this up
